### PR TITLE
Add Cognito Authenticator implementation

### PR DESCRIPTION
**Pull request by senthalan**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -21,3 +27,8 @@
 
 # virtual machine crash logs, see http://www.java.com/en/download/help/error_hotspot.xml
 hs_err_pid*
+
+
+# Ignore everything in this directory
+target
+.idea
```
Comment (mefarazath) [inline .gitignore]: Add a new line

File: components/org.wso2.carbon.identity.application.authenticator.cognito/pom.xml [hunk 2]
```diff
@@ -0,0 +1,112 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+  ~ Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/pom.xml]: Check Licence year

File: components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+    private static Log log = LogFactory.getLog(CognitoOIDCAuthenticator.class);
+    private String tokenEndpoint;
+    private String oAuthEndpoint;
+    private String logoutEndpoint;
+    private String userInfoEnoint;
+
+    @Override
+    public boolean canHandle(HttpServletRequest request) {
+
+        if (log.isTraceEnabled()) {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java]: trace or debug?

File: components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/internal/CognitoOpenIDConnectAuthenticatorServiceComponent.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    private static Log log = LogFactory.getLog(CognitoOpenIDConnectAuthenticatorServiceComponent.class);
+
+    @Activate
+    protected void activate(ComponentContext ctxt) {
+        try {
+            CognitoOIDCAuthenticator cognitoOIDCAuthenticator = new CognitoOIDCAuthenticator();
+            ctxt.getBundleContext().registerService(ApplicationAuthenticator.class.getName(),
+                    cognitoOIDCAuthenticator, null);
+            if (log.isDebugEnabled()) {
+                log.debug("OpenID Connect Authenticator bundle is activated");
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/internal/CognitoOpenIDConnectAuthenticatorServiceComponent.java]: OpenID Connect Authenticator --> Cognito Authenticator?

File: components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/internal/CognitoOpenIDConnectAuthenticatorServiceComponent.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            }
+        } catch (Throwable e) {
+            log.fatal(" Error while activating oidc authenticator ", e);
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/internal/CognitoOpenIDConnectAuthenticatorServiceComponent.java]: OIDC Authenticator --> Cognito Authenticator?

File: components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/internal/CognitoOpenIDConnectAuthenticatorServiceComponent.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        }
+    }
+
+    @Deactivate
+    protected void deactivate(ComponentContext ctxt) {
+        if (log.isDebugEnabled()) {
+            log.debug("OpenID Connect Authenticator bundle is deactivated");
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/internal/CognitoOpenIDConnectAuthenticatorServiceComponent.java]: OIDC Authenticator --> Cognito Authenticator?

File: components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+        scope.setDisplayName("Additional Query Parameters");
+        scope.setName(CognitoOIDCAuthenticatorConstants.ADDITIONAL_QUERY_PARAMS);
+        scope.setDescription("Additional query parameters. e.g: paramName1=value1");
+        scope.setDisplayOrder(6);
+        configProperties.add(scope);
+
+        return configProperties;
+    }
+
+//    @Override
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java]: Remove commented code

File: components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+    }
+
+    @Override
+    public String getContextIdentifier(HttpServletRequest request) {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Inside CognitoOIDCAuthenticator.getContextIdentifier()");
+        }
+        Cookie cookie = FrameworkUtils.getCookie(request, CognitoOIDCAuthenticatorConstants
+                .COGNITO_LOGOUT_STATE_COOKIE);
```
Comment (Kavindu-Dodan) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java]: Can improve formatting. Take function call to a new line

File: components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticatorConstants.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+    public static final String COGNITO_LOGOUT_CLIENT_ID = "client_id";
+    public static final String COGNITO_LOGOUT_URI = "logout_uri";
+    public static final String COGNITO_LOGOUT_STATE_COOKIE = "cognito-logout-state";
+
+    public static final String COGNITO_OAUTH_URL = "/oauth2/authorize";
+    public static final String COGNITO_TOKEN_URL = "/oauth2/token";
+    public static final String COGNITO_USER_INFO__URL = "/oauth2/userInfo";
+    public static final String COGNITO_LOGOUT__URL = "/logout";
+}
```
Comment (Kavindu-Dodan) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticatorConstants.java]: New line

File: components/org.wso2.carbon.identity.application.authenticator.cognito/pom.xml [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+                            org.wso2.carbon.identity.application.authenticator.oidc;
+                            version="${carbon.identity.authenticator.oidc.imp.pkg.version.range}",
+                            org.wso2.carbon.identity.core.util;
+                            version="${carbon.identity.framework.package.import.version.range}",
+                        </Import-Package>
+                        <Export-Package>
+                            !org.wso2.carbon.identity.application.authenticator.cognito.internal,
+                            org.wso2.carbon.identity.application.authenticator.cognito*
+                        </Export-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/pom.xml]: Do we need dynamic imports?

File: components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+        Map<String, String> parameters = new HashMap<>();
+
+        parameters.put(CognitoOIDCAuthenticatorConstants.COGNITO_LOGOUT_CLIENT_ID, clientId);
+        parameters.put(CognitoOIDCAuthenticatorConstants.COGNITO_LOGOUT_URI, logoutRedirectUrl);
+        FrameworkUtils.setCookie(request, response, CognitoOIDCAuthenticatorConstants.COGNITO_LOGOUT_STATE_COOKIE,
+                context.getContextIdentifier(), null);
+        try {
+            response.sendRedirect(logoutUrl + "?" + getParamsString(parameters));
+        } catch (IOException e) {
+            throw new LogoutFailedException("Error while triggering cognito logout", e);
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java]: add some context info, eg: client_id, the logout URL

File: components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+                .LOGOUT_REDIRECT_URL);
+
+        Map<String, String> parameters = new HashMap<>();
+
+        parameters.put(CognitoOIDCAuthenticatorConstants.COGNITO_LOGOUT_CLIENT_ID, clientId);
+        parameters.put(CognitoOIDCAuthenticatorConstants.COGNITO_LOGOUT_URI, logoutRedirectUrl);
+        FrameworkUtils.setCookie(request, response, CognitoOIDCAuthenticatorConstants.COGNITO_LOGOUT_STATE_COOKIE,
+                context.getContextIdentifier(), null);
+        try {
+            response.sendRedirect(logoutUrl + "?" + getParamsString(parameters));
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java]: logoutUrl + "?" + getParamsString(parameters)

Can we check whether there are available Util methods to achieve the same in FrameworkUtils class?

File: pom.xml [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-deploy-plugin</artifactId>
+            </plugin>
+            <plugin>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>${maven.compiler.plugin.version}</version>
+                <inherited>true</inherited>
+                <configuration>
+                    <encoding>UTF-8</encoding>
+                    <source>1.7</source>
```
Comment (mefarazath) [inline pom.xml]: 1.8?

File: pom.xml [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+  ~  specific language governing permissions and limitations
+  ~  under the License.
+  -->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+    <parent>
+        <groupId>org.wso2</groupId>
+        <artifactId>wso2</artifactId>
+        <version>1</version>
```
Comment (mefarazath) [inline pom.xml]: Shall we try to use v5 pom if possible?
That will allow us to incorporate checkstyle and other best practices easily as they are validated during build time.

File: components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+    private String logoutEndpoint;
+    private String userInfoEnoint;
+
+    @Override
+    public boolean canHandle(HttpServletRequest request) {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Inside CognitoOIDCAuthenticator.canHandle()");
+        }
+        boolean canHandle = super.canHandle(request);
```
Comment (Kavindu-Dodan) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java]: simply return this at the end of the function. canHandle variable is not used

File: components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java [hunk 16]
```diff
@@ -0,0 +1,319 @@
...
+
+        try {
+            for (Map.Entry<String, String> entry : params.entrySet()) {
+                result.append(URLEncoder.encode(entry.getKey(), "UTF-8"));
+                result.append("=");
+                result.append(URLEncoder.encode(entry.getValue(), "UTF-8"));
+                result.append("&");
+            }
+
+            String resultString = result.toString();
```
Comment (Kavindu-Dodan) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java]: try catch is only required for for loop. so the return can be taken out of it which will make things compact and clean

File: components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Authenticator to handle log in ang log out of the AWS Cognito.
+ */
+public class CognitoOIDCAuthenticator extends OpenIDConnectAuthenticator {
+
+    private static final long serialVersionUID = 9058607724358986002L;
+    private static Log log = LogFactory.getLog(CognitoOIDCAuthenticator.class);
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java]: private static final

File: components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java [hunk 18]
```diff
@@ -0,0 +1,294 @@
...
+    private String tokenEndpoint;
+    private String oAuthEndpoint;
+    private String logoutEndpoint;
+    private String userInfoEndpoint;
+
+    @Override
+    public boolean canHandle(HttpServletRequest request) {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Inside CognitoOIDCAuthenticator.canHandle()");
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java]: Log line like Inside CognitoOIDCAuthenticator.canHandle() looms primitive.
instead use
"Evaluating if request can be handled"
Also can add these into the context
request.getParameter
                (CognitoOIDCAuthenticatorConstants.COGNITO_STATE)) 
getContextIdentifier(request)

File: components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java [hunk 19]
```diff
@@ -0,0 +1,294 @@
...
+        String claimDialectUri = super.getClaimDialectURI();
+        AuthenticatorConfig authConfig = FileBasedConfigurationBuilder.getInstance().getAuthenticatorBean(getName());
+        if (authConfig != null) {
+            Map<String, String> parameters = authConfig.getParameterMap();
+            if (parameters != null && parameters.containsKey(CognitoOIDCAuthenticatorConstants.
+                    CLAIM_DIALECT_URI_PARAMETER)) {
+                claimDialectUri = parameters.get(CognitoOIDCAuthenticatorConstants.CLAIM_DIALECT_URI_PARAMETER);
+            } else {
+                if (log.isDebugEnabled()) {
+                    log.debug("Found no Parameter map for connector " + getName());
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java]: log.debug("Found no Parameter map for connector  : " + getName());

note ":", 
Please have a separator for any context param value for all logs

File: components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java [hunk 20]
```diff
@@ -0,0 +1,294 @@
...
+        }
+        if (log.isDebugEnabled()) {
+            log.debug("Authenticator " + getName() + " is using the claim dialect uri " + claimDialectUri);
+        }
+        return claimDialectUri;
+    }
+
+    private void initTokenEndpoint() {
+
+        this.tokenEndpoint = getAuthenticatorConfig().getParameterMap().get(CognitoOIDCAuthenticatorConstants
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java]: getAuthenticatorConfig().getParameterMap() can this be null ?

File: components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java [hunk 21]
```diff
@@ -0,0 +1,294 @@
...
+
+        return configProperties;
+    }
+
+    @Override
+    protected void initiateLogoutRequest(HttpServletRequest request, HttpServletResponse response,
+                                         AuthenticationContext context) throws LogoutFailedException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Sending logout request to external IDP");
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/src/main/java/org/wso2/carbon/identity/application/authenticator/cognito/CognitoOIDCAuthenticator.java]: better to have a context info in log, eg. IDP URL .

**Review by ruwanta — COMMENTED** [review]
Few improvements on logs

**Review by ruwanta — CHANGES_REQUESTED** [review]
Need unit tests and line coverage of 80%

File: components/org.wso2.carbon.identity.application.authenticator.cognito/pom.xml [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+
+                            org.apache.oltu.oauth2.client; version="${oltu.package.import.version.range}",
+                            org.apache.oltu.oauth2.client.request; version="${oltu.package.import.version.range}",
+                            org.apache.oltu.oauth2.client.response; version="${oltu.package.import.version.range}",
+                            org.apache.oltu.oauth2.common.exception; version="${oltu.package.import.version.range}",
+                            org.apache.oltu.oauth2.common.message.types;
+                            version="${oltu.package.import.version.range}",
+                            org.apache.oltu.oauth2.common.utils; version="${oltu.package.import.version.range}",
+                            org.wso2.carbon.utils.*; version="${carbon.kernel.package.import.version.range}",
+                            net.minidev.json; version="${net.minidev.json.imp.pkg.version.range}",
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/pom.xml]: is minidev.json needed here ?

Comment (senthalan) [reply components/org.wso2.carbon.identity.application.authenticator.cognito/pom.xml]: It is used in OpenIDConnectAuthenticator https://github.com/wso2-extensions/identity-outbound-auth-oidc/blob/master/components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java#L20

Comment (ruwanta) [reply components/org.wso2.carbon.identity.application.authenticator.cognito/pom.xml]: Which means we do not need it here

File: components/org.wso2.carbon.identity.application.authenticator.cognito/pom.xml [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.powermock</groupId>
+            <artifactId>powermock-module-junit4</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon.identity.framework</groupId>
+            <artifactId>org.wso2.carbon.identity.testutil</artifactId>
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/pom.xml:78]: pls add <scope>test</scope>

File: components/org.wso2.carbon.identity.application.authenticator.cognito/pom.xml [hunk 24]
```diff
@@ -0,0 +1,189 @@
...
+                                <exclude>**/CognitoOIDCAuthenticatorConstants*</exclude>
+                            </excludes>
+                            <rules>
+                                <rule implementation="org.jacoco.maven.RuleConfiguration">
+                                    <element>BUNDLE</element>
+                                    <limits>
+                                        <limit implementation="org.jacoco.report.check.Limit">
+                                            <counter>COMPLEXITY</counter>
+                                            <value>COVEREDRATIO</value>
+                                            <minimum>0.45</minimum>
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.application.authenticator.cognito/pom.xml]: any reason this is set to too low? we need at least 60% without debug log lines

**Review by ruwanta — APPROVED** [review]

**Pull request merged**